[docs] Refactor `Localization` documentation sections (#11989)
mui/mui-x: docs/data/date-pickers/date-field/date-field.md

Learn more about the _Controlled and uncontrolled_ pattern in the [React documentation](https://react.dev/learn/sharing-state-between-components#controlled-and-uncontrolled-components).
:::

## Localization

See the [Date format and localization](/x/react-date-pickers/adapters-locale/) and [Translated components](/x/react-date-pickers/localization/) documentation pages for more details.

### Customize the date format

{{"demo": "CustomDateFormat.js"}}

:::info

> You can find more information about [custom formats](/x/react-date-pickers/adapters-locale/#custom-formats) in the documentation page.

Reviewer: From what I see most of the time we use the following structure to refer to another doc page:

```suggestion
You can find more information in the [Date format and localization page](/x/react-date-pickers/adapters-locale/#custom-formats)
```

Or 

```suggestion
You can find more information in the [Date format and localization](/x/react-date-pickers/adapters-locale/#custom-formats) documentation page
```

It would be nice to be more consistent in the whole doc of course, @samuelsycamore if you have a preference :smiley:

Author: Or:
```suggestion
See [custom formats](/x/react-date-pickers/adapters-locale/#custom-formats) documentation for more information.
```
Avoid specifying `page`, when we are talking about a section. 👌

Reviewer: True, if often use "See" as well
We even have some "Take a look"

Author: There is no rush for this, let's wait for Sam's input. 👌

Author: Thank you @samuelsycamore. 🙏 
I've changed it to `Date format and localization—Custom formats` instead of `Date localization—Custom formats` to correctly reflect the page name.

{{"demo": "CustomDateFormat.js"}}

> :::info

Reviewer: IMHO this demo should either not exist or be inside the "Localization" section below
THis would help people understand that the custom date format is part of the the Localization in our doc

Author: Initially, I was leaning towards that direction, but then I thought that the various custom formats for different components (date, time, and date-time) have lots of examples. If we want to provide at least a similar amount of demos in the `Custom format` section, it would be huge. 🤯 
That's why I decided to keep the separate section. 🤔 
WDYT?
Would moving those `Customize the <x> format` sections as sub-sections of `Localization` make more sense to you?

Author: @flaviendelangle I've moved the `Customize the format` sections as subsections of `Localization`.
Let me know what you think. 😉

Reviewer: I can be a good trade-off :ok_hand:

Reviewer 2: This can be a great use-case for a switch like github does it on their webhook documentation:
<img width="801" alt="Screenshot 2024-02-08 at 14 20 15" src="https://github.com/mui/mui-x/assets/32863416/aaec8f8c-413e-457a-bbec-f3cc3ca3eae4">

Reviewer 2: In the sense that you could provide a few options to select formats and then the fields will use that. WDYT?

Author: That is a great suggestion, I'll explore that option. 👌
